### PR TITLE
feat: Record Advance Payment as a liability

### PR DESCRIPTION
**Pull request by GursheenK**
The current flow allows the user to record - 

- **Advance Payments Received** from Customers as a **Negative Asset** in the General Ledger
- **Advance Payments Paid** to Suppliers as a **Negative Liability** in the General Ledger

The **Book Advance Payments in Separate Party Account** feature introduces the option to use separate accounts for recording these Advance Payments. The option can be selected from the **Company Doctype** > **Advance Payments Section**

When selected, the user needs to provide the Default Advances Received Account and the Default Advances Paid Account in the Company master.
<br>
Screenshot 2023-11-13 at 4 10 57 PM
<br>
Default Advance Accounts can be set for a Customer/Customer Group/Supplier/ Supplier Group with respect to multiple companies from the **Accounting Tab** > **Default Accounts Table**
<br>
Screenshot 2023-11-13 at 4 14 32 PM
<br>

> Note : If they are not set in the Customer or Supplier Doctypes, then the Default Advance Account mentioned in the Company Master will be used for recording the Advance Payments of that particular party.

### **Example**

A Payment Entry for receiving `₹1000` from a customer is submitted. The General Ledger Entries created for the same will use the Liability Account instead of the Receivable Account since it is an Advance Payment. 
<br>
>Screenshot 2023-11-13 at 4 16 47 PM

<br>

Suppose a Sales Invoice with a total of `₹500` is created for the same customer and the Advance Payments are fetched and allocated using the _Get Advances Received_ button. In such a case, when the Sales Invoice is submitted, the General Ledger for the Payment Entry will reflect the following changes.

<br>

>Screenshot 2023-11-13 at 4 19 54 PM

<br>

Since the total allocated amount to the Sales Invoice is `₹500`, the same has been debited from the Liability Account. The Liability Account will then hold the remaining `₹500` as an Advance Payment.

<br>

`no-docs`

Issue #34282

### CODE REVIEW

**Comment by s-aga-r** [comment]
@deepeshgarg007 @GursheenK Why two [PRs](https://github.com/frappe/erpnext/pull/35537)?

File: erpnext/accounts/doctype/payment_entry/payment_entry.py [hunk 1]
```diff
@@ -86,12 +86,36 @@ def validate(self):
 	def on_submit(self):
 		if self.difference_amount:
 			frappe.throw(_("Difference Amount must be zero"))
+		book_advance_payments_as_liability = frappe.get_value(
+			"Company", {"company_name": self.company}, "book_advance_payments_as_liability"
+		)
+		if book_advance_payments_as_liability:
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/payment_entry/payment_entry.py]: Simply call `get_liability_account` from here and check if `book_advance_payments_as_liability` is enabled or not inside the function itself

File: erpnext/accounts/doctype/payment_entry/payment_entry.py [hunk 2]
```diff
@@ -86,12 +86,36 @@ def validate(self):
 	def on_submit(self):
 		if self.difference_amount:
 			frappe.throw(_("Difference Amount must be zero"))
+		book_advance_payments_as_liability = frappe.get_value(
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/payment_entry/payment_entry.py]: This should be on `validate`, not on submit

File: erpnext/accounts/doctype/payment_entry/payment_entry.js [hunk 3]
```diff
@@ -708,7 +712,7 @@ frappe.ui.form.on('Payment Entry', {
 				if(r.message) {
 					var total_positive_outstanding = 0;
 					var total_negative_outstanding = 0;
-
+					console.log(r.message);
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/payment_entry/payment_entry.js]: Remove this

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]

**Comment by rtdany10** [comment]
~@deepeshgarg007 will this be backported?~
Just saw the label, but any possibility?

**Pull request merged**